### ISSUE TITLE
BuildStream crashes on junctions in non-strict mode

### ISSUE THREAD

**Issue by Cynical-Optimist** [buildstream-migration/bst-staging#1018]
[See original issue on GitLab](https://gitlab.com/BuildStream/buildstream/-/issues/1018)
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on May 8, 2019, 12:05

## Summary

BuildStream crashes on junctions in non-strict mode

## Steps to reproduce

Build or fetch any project which has a junction, in non-strict mode.

## What is the current bug behavior?

```
Traceback (most recent call last):
  File "/codethink/GNOME/buildstream/buildstream/testing/runcli.py", line 393, in _invoke
    cli_object.main(args=args or (), prog_name=cli_object.name)
  File "/codethink/GNOME/buildstream/buildstream/_frontend/cli.py", line 195, in override_main
    standalone_mode=standalone_mode, **extra)
  File "/codethink/GNOME/buildstream/.tox/py36/lib/python3.6/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/codethink/GNOME/buildstream/.tox/py36/lib/python3.6/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/codethink/GNOME/buildstream/.tox/py36/lib/python3.6/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/codethink/GNOME/buildstream/.tox/py36/lib/python3.6/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/codethink/GNOME/buildstream/.tox/py36/lib/python3.6/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/codethink/GNOME/buildstream/.tox/py36/lib/python3.6/site-packages/click/decorators.py", line 27, in new_func
    return f(get_current_context().obj, *args, **kwargs)
  File "/codethink/GNOME/buildstream/buildstream/_frontend/cli.py", line 672, in source_fetch
    remote=remote)
  File "/codethink/GNOME/buildstream/buildstream/_stream.py", line 324, in fetch
    source_remote_url=remote)
  File "/codethink/GNOME/buildstream/buildstream/_stream.py", line 1061, in _load
    fetch_subprojects=fetch_subprojects)
  File "/codethink/GNOME/buildstream/buildstream/_pipeline.py", line 113, in load
    fetch_subprojects=fetch_subprojects)
  File "/codethink/GNOME/buildstream/buildstream/_project.py", line 455, in load_elements
    fetch_subprojects=fetch_subprojects)
  File "/codethink/GNOME/buildstream/buildstream/_loader/loader.py", line 114, in load
    element = loader._load_file(name, rewritable, ticker, fetch_subprojects)
  File "/codethink/GNOME/buildstream/buildstream/_loader/loader.py", line 258, in _load_file
    fetch_subprojects=fetch_subprojects, provenance=dep.provenance)
  File "/codethink/GNOME/buildstream/buildstream/_loader/loader.py", line 569, in _get_loader
    filename, element._get_cache_key())
  File "/usr/local/lib/python3.6/posixpath.py", line 94, in join
    genericpath._check_arg_types('join', a, *p)
  File "/usr/local/lib/python3.6/genericpath.py", line 149, in _check_arg_types
    (funcname, s.__class__.__name__)) from None
TypeError: join() argument must be str or bytes, not 'NoneType'

```

## What is the expected correct behavior?

No crashes please, just happily load the junction cache key.

## Additional information

This only happens with `master`, not in `bst-1.2`.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on May 8, 2019, 12:11

mentioned in commit 5a53b1912c44cf11ec4eb51652dab3c646e80a5b

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on May 8, 2019, 12:11

mentioned in commit 9dcc97da95d947dc68f938b506f52c8d0e6c1607

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on May 8, 2019, 12:12

mentioned in merge request !1333

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on May 8, 2019, 12:13

mentioned in commit a0e2b07bc52cb327c0398432427298e91ea5d6fb

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on May 8, 2019, 12:13

mentioned in commit 26945b53bdf4b00acd16f15e64d35fb0154cc928

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on May 8, 2019, 12:42

mentioned in commit 000ae7beabc0f85230f760a506f5f53779aa2701

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on May 8, 2019, 12:42

mentioned in commit 34c3034b959ff2b839f17a7cbf998f95aca5ff58

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on May 8, 2019, 13:09

marked #993 as a duplicate of this issue

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on May 9, 2019, 07:31

closed via merge request !1333

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on May 9, 2019, 07:31

mentioned in commit 9800accc7432c7ac3f375d7fa0c5f8ed28cd47ff

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on May 9, 2019, 07:42

Looks like my comment yesterday got lost somewhere between my browser cache and gitlab.com, so here goes again:

I've closed the original #993 as duplicate of this issue since I didn't notice it until setting up the MRs to fix this one, but it's worth noting [[Gitlab user @abderrahimk]](https://gitlab.com/abderrahimk)'s original findings:

> Bisecting this gives it started happening in 352f4ad9. It seems `element._get_cache_key()` is returning `None` for some reason.
